File: 22-Strategy/src/go/credit_card_strategy.go
package main

import "fmt"

// CreditCardStrategy is a Concrete Strategy — pays with a credit card.
type CreditCardStrategy struct {
	cardNumber string
}

func (c *CreditCardStrategy) Pay(amount float64) {
	masked := "****" + c.cardNumber[len(c.cardNumber)-4:]
	fmt.Printf("  Paid $%.2f with credit card %s\n", amount, masked)
}

File: 22-Strategy/src/rust/credit_card_strategy.rs
use crate::payment_strategy::PaymentStrategy;

/// Concrete Strategy — pays with a credit card.
pub struct CreditCardStrategy {
    card_number: String,
}

impl CreditCardStrategy {
    pub fn new(card_number: &str) -> Self {
        CreditCardStrategy { card_number: card_number.to_string() }
    }
}

impl PaymentStrategy for CreditCardStrategy {
    fn pay(&self, amount: f64) {
        let masked = format!("****{}", &self.card_number[self.card_number.len() - 4..]);
        println!("  Paid ${:.2} with credit card {}", amount, masked);
    }
}
